### PR TITLE
control checking out head instead of base ref by a GH label

### PR DESCRIPTION
**Pull request by hghaf099**
Addresses https://hashicorp.atlassian.net/browse/VAULT-19349
The label only affects checkouts for tests now, and I think that is good enough since we should not be merging a PR into a broken main.

Context from the ticket:
This blog post covers all the gory details: [What is a GitHub Pull Request merge branch?](https://fluffyandflakey.blog/2022/12/21/what-is-a-github-pull-request-merge-branch) The TL;DR is that when your PR is checked out, it's not just your branch that's used. They do a sort of implicit rebase against the parent branch.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -64,11 +64,25 @@ jobs:
...
+    runs-on: ubuntu-latest
+    outputs:
+      checkout-ref: ${{ steps.checkout-ref-output.outputs.checkout-ref }}
+    steps:
+      - if: ${{ !contains(github.event.pull_request.labels.*.name, 'checkout-head') }}
+        run: echo "CHECKOUT_REF=${{ github.base_ref }}" >> "$GITHUB_ENV"
+      - if: ${{ contains(github.event.pull_request.labels.*.name, 'checkout-head') }}
+        run: echo "CHECKOUT_REF=${{ github.event.pull_request.head.sha }}" >> "$GITHUB_ENV"
+      - id: checkout-ref-output
+        run: echo "checkout-ref=${{ env.CHECKOUT_REF }}" >> "$GITHUB_OUTPUT"
```
Comment (kubawi) [inline .github/workflows/ci.yml:78]: Could we move this logic into the `Setup` job above? Each job runs on a separate runner, so we can save ourselves acquiring a `ubuntu-latest` runner each CI run if it's part of the setup logic 😄

**Review by kubawi — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (hghaf099) [reply .github/workflows/ci.yml:68]: Good call! Done

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -59,6 +60,12 @@ jobs:
         with:
           github-token: ${{ secrets.ELEVATED_GITHUB_TOKEN }}
           no-restore: true # don't download them on a cache hit
+      - if: ${{ !contains(github.event.pull_request.labels.*.name, 'checkout-head') }}
```
Comment (kubawi) [inline .github/workflows/ci.yml:63]: Could you please add either `name:` fields or comments explaining why we're doing this? Just to save someone figuring it out months from now :D

**Review by kubawi — APPROVED** [review]
Looks good! I've left one minor comment.

**Pull request merged**